File: 13. CombSort.py
def comb_sort(arr):
    n = len(arr)
    gap = n
    shrink = 1.3
    sorted = False

    while not sorted:
        # Update the gap value
        gap = int(gap / shrink)
        if gap <= 1:
            gap = 1
            sorted = True  # Assume the array is sorted

        # Compare elements with the current gap
        for i in range(n - gap):
            if arr[i] > arr[i + gap]:
                # Swap elements
                arr[i], arr[i + gap] = arr[i + gap], arr[i]
                sorted = False  # If a swap happens, the array is not sorted yet

        print(f"Array after gap {gap}: {' '.join(map(str, arr))}")

def main():
    """Main function to test the comb_sort."""
    arr = [8, 4, 1, 56, 3, -44, 23, -6, 28, 0]
    
    print("Original array: ", ' '.join(map(str, arr)))
    
    comb_sort(arr)
    
    print("Sorted array (Comb Sort): ", ' '.join(map(str, arr)))

if __name__ == "__main__":
    main()
    """
    Sorts a list in place using the Comb Sort algorithm.

    Type: Comparison-based, Not stable, In-place

    Time Complexity:
    Best Case: O(n log n) -> When the array is nearly sorted
    Average Case: O(n^2 / 2^p) where p is the number of increments. Roughly O(n log n).
    Worst Case: O(n^2) -> Though rare with the 1.3 shrink factor.

    Space Complexity:
    Auxiliary Space: O(1) -> In-place sorting
    """
    

File: 13. CombSort.cpp
#include <bits/stdc++.h>
using namespace std;

void combSort(vector<int> &arr) {
    int n = arr.size();
    int gap = n;
    const double shrink = 1.3;
    bool sorted = false;

    while(!sorted) {
        gap = int(gap / shrink);
        if(gap <= 1) {
            gap = 1;
            sorted = true;
        }

        for(int i=0; i+gap<n; i++) {
            if(arr[i] > arr[i+gap]) {
                swap(arr[i], arr[i+gap]);
                sorted = false;
            }
        }
        cout << "Array after gap " << gap << ": ";
        for(int num : arr) cout << num << " ";
        cout << endl;
    }
}

int main() {
    vector<int> arr = {8, 4, 1, 56, 3, -44, 23, -6, 28, 0};
    cout << "Original array: ";
    for(int num : arr) cout << num << " ";
    cout << endl;

    combSort(arr);

    cout << "Sorted array (Comb Sort): ";
    for(int num : arr) cout << num << " ";
    cout << endl;
    return 0;
}

/*
Type: Comparison-based, Not stable, In-place

Time Complexity:
Best Case: O(n log n) → When array is nearly sorted
Average Case: O(n² / 2^p) ≈ O(n²) → Depends on shrink factor and gap sequence
Worst Case: O(n²) → Reverse sorted array or bad gap sequence

Space Complexity:
Auxiliary Space: O(1) → In-place sorting

*/